text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,972 @@
+lifetimes that are bound within the type we are looking at.  Let's
+call that set `R=<r0..rn>`. Initially, this set `R` is empty, but it
+will grow as we traverse through types like fns or objects, which can
+bind region names.
+
+#### Simple outlives rules
+
+Here are the rules covering the simple cases, where no type parameters
+or projections are involved:
+
+    OutlivesScalar:
+      --------------------------------------------------
+      R ⊢ scalar: 'a
+
+    OutlivesNominalType:
+      ∀i. Pi: 'a
+      --------------------------------------------------
+      R ⊢ Id<P0..Pn>: 'a
+
+    OutlivesReference:
+      R ⊢ 'x: 'a
+      R ⊢ T: 'a
+      --------------------------------------------------
+      R ⊢ &'x T: 'a
+
+    OutlivesObject:
+      ∀i. R ⊢ Oi: 'a
+      R ⊢ 'x: 'a
+      --------------------------------------------------
+      R ⊢ O0..On+'x: 'a
+
+    OutlivesFunction:
+      ∀i. R,r.. ⊢ Ti: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0
+
+    OutlivesTraitRef:
+      ∀i. R,r.. ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> TraitId<P0..Pn>: 'a      
+      
+#### Outlives for lifetimes
+
+The outlives relation for lifetimes depends on whether the lifetime in
+question was bound within a type or not. In the usual case, we decide
+the relationship between two lifetimes by consulting the environment.
+Lifetimes representing scopes within the current fn have a
+relationship derived from the code itself, lifetime parameters have
+relationships defined by where-clauses and implied bounds:
+
+      'x not in R
+      ('x: 'a) in Env
+      --------------------------------------------------
+      R ⊢ 'x: 'a
+      
+For higher-ranked lifetimes, we simply ignore the relation, since the
+lifetime is not yet known. This means for example that `fn<'a> fn(&'a
+i32): 'x` holds, even though we do not yet know what region `'a` is
+(and in fact it may be instantiated many times with different values
+on each call to the fn).
+      
+      'x in R
+      --------------------------------------------------
+      R ⊢ 'x: 'a
+
+#### Outlives for type parameters
+
+For type parameters, the only way to draw "outlives" conclusions is to
+find information in the environment (which is being threaded
+implicitly here, since it is never modified). In terms of a Rust
+program, this means both explicit where-clauses and implied bounds
+derived from the signature (discussed below).
+
+    OutlivesTypeParameterEnv:
+      X: 'a in Env
+      --------------------------------------------------
+      R ⊢ X: 'a
+
+
+#### Outlives for projections
+
+Projections have the most possibilities. First, we may find
+information in the environment, as with type parameters, but we can
+also consult the trait definition to find bounds (consider an
+associated type declared like `type Foo: 'static`). These rule only
```

Comment:
It's interesting to me that trait definitions are not considered part of the environment when talking about a type that implements that trait.